"test_UoI_NMF_fit_tranform" test sometimes failes on python 3.5

```
__________________________ test_UoI_NMF_fit_tranform ___________________________

    @pytest.mark.fast
    def test_UoI_NMF_fit_tranform():
        """Tests that the """
        uoi, X = setup()
>       uoi.fit(X)
```

```
        if ensure_min_samples > 0:
            n_samples = _num_samples(array)
            if n_samples < ensure_min_samples:
                raise ValueError("Found array with %d sample(s) (shape=%s) while a"
                                 " minimum of %d is required%s."
                                 % (n_samples, array.shape, ensure_min_samples,
>                                   context))

E               ValueError: Found array with 0 sample(s) (shape=(0, 2)) while a minimum of 1 is required by the normalize function.
```

--- a/pyuoi/decomposition/NMF.py
+++ b/pyuoi/decomposition/NMF.py
@@ -3,9 +3,11 @@ import numpy as np
 import logging
 
 from .base import AbstractDecompositionModel
+from .utils import dissimilarity
 
 from ..utils import check_logger
 
+from itertools import combinations
 from sklearn.decomposition import NMF as skNMF
 from sklearn.cluster import DBSCAN
 from sklearn.preprocessing import normalize
@@ -38,6 +40,10 @@ class UoI_NMF_Base(AbstractDecompositionModel):
     cons_meth : function
         The method for computing consensus bases after clustering. If None,
         uses np.median.
+    use_dissimilarity : bool
+        Whether to use dissimilarity to choose the final rank. If False, all
+        bases across ranks are concatenated and clustered. The final rank in
+        this case is how many clusters are chosen.
     random_state : int, RandomState instance, or None
         The seed of the pseudo random number generator that selects a random
         feature to update.  If int, random_state is the seed used by the random
@@ -51,7 +57,7 @@ class UoI_NMF_Base(AbstractDecompositionModel):
     """
     def __init__(
         self, n_boots=10, ranks=None, nmf=None, cluster=None, nnreg=None,
-        cons_meth=None, random_state=None, logger=None
+        cons_meth=None, use_dissimilarity=True, random_state=None, logger=None
     ):
         self.__initialize(
             n_boots=n_boots,
@@ -61,8 +67,7 @@ class UoI_NMF_Base(AbstractDecompositionModel):
             nnreg=nnreg,
             cons_meth=cons_meth,
             logger=logger,
-            random_state=random_state
-        )
+            random_state=random_state)
 
     def set_params(self, **kwargs):
         """Set the parameters of this estimator."""
@@ -163,25 +168,41 @@ class UoI_NMF_Base(AbstractDecompositionModel):
         check_non_negative(X, 'UoI NMF')
         n_samples, n_features = X.shape
 
-        k_tot = sum(self.ranks)
-        n_H_samples = k_tot * self.n_boots
-        H_samples = np.zeros((n_H_samples, n_features))
+        H_samples = {k: np.zeros((self.n_boots, k, n_features))
+                     for k in self.ranks}
 
         rep_idx = self._rand.randint(n_samples, size=(self.n_boots, n_samples))
-        for i in range(self.n_boots):
-            self._logger.info("bootstrap %d" % i)
+        for boot_idx in range(self.n_boots):
+            self._logger.info("Bootstrap %d" % boot_idx)
             # compute NMF bases for k across bootstrap replicates
-            H_i = i * k_tot
-            sample = X[rep_idx[i]]
+            sample = X[rep_idx[boot_idx]]
             for k_idx, k in enumerate(self.ranks):
                 # concatenate k by p
-                H_samples[H_i:H_i + k:, ] = (self.nmf.set_params(n_components=k)
-                                             .fit(sample).components_)
-                H_i += k
+                H_samples[k][boot_idx] = \
+                    self.nmf.set_params(n_components=k).fit(sample).components_
+
+        if self.use_dissimilarity:
+            gamma = np.zeros(self.ranks.size)
+            for k_idx, k in enumerate(self.ranks):
+                H_k = H_samples[k]
+                for boot1, boot2 in combinations(range(self.n_boots), 2):
+                    gamma[k_idx] += dissimilarity(H_k[boot1], H_k[boot2])
+            k_min = self.ranks[np.argmin(gamma)]
+            H_pre_cluster = H_samples[k_min].reshape((self.n_boots * k,
+                                                      n_features))
+        else:
+            H_pre_cluster = np.zeros((self.n_boots * np.sum(self.ranks),
+                                      n_features))
+            start_idx = 0
+            for k in self.ranks:
+                H_k = H_samples[k].reshape((self.n_boots * k, n_features))
+                end_idx = start_idx + self.n_boots * k
+                H_pre_cluster[start_idx:end_idx] = H_k
+                start_idx = end_idx
 
         # remove zero bases and normalize across features
-        H_samples = H_samples[np.sum(H_samples, axis=1) != 0.0]
-        H_samples = normalize(H_samples, norm='l2', axis=1)
+        H_pre_cluster = H_pre_cluster[np.sum(H_samples, axis=1) != 0.0]
+        H_pre_cluster = normalize(H_pre_cluster, norm='l2', axis=1)
 
         # cluster all bases
         self._logger.info("clustering bases samples")

--- a/pyuoi/decomposition/NMF.py
+++ b/pyuoi/decomposition/NMF.py
@@ -68,6 +68,7 @@ class UoI_NMF_Base(AbstractDecompositionModel):
             cons_meth=cons_meth,
             logger=logger,
             random_state=random_state)
+        self.use_dissimilarity = use_dissimilarity
 
     def set_params(self, **kwargs):
         """Set the parameters of this estimator."""
@@ -182,13 +183,17 @@ class UoI_NMF_Base(AbstractDecompositionModel):
                     self.nmf.set_params(n_components=k).fit(sample).components_
 
         if self.use_dissimilarity:
-            gamma = np.zeros(self.ranks.size)
+            gamma = np.zeros(len(self.ranks))
+            # iterate over each rank
             for k_idx, k in enumerate(self.ranks):
+                # extract the bases for each rank
                 H_k = H_samples[k]
                 for boot1, boot2 in combinations(range(self.n_boots), 2):
                     gamma[k_idx] += dissimilarity(H_k[boot1], H_k[boot2])
             k_min = self.ranks[np.argmin(gamma)]
-            H_pre_cluster = H_samples[k_min].reshape((self.n_boots * k,
+            print(gamma)
+            print(k_min)
+            H_pre_cluster = H_samples[k_min].reshape((self.n_boots * k_min,
                                                       n_features))
         else:
             H_pre_cluster = np.zeros((self.n_boots * np.sum(self.ranks),
@@ -201,12 +206,12 @@ class UoI_NMF_Base(AbstractDecompositionModel):
                 start_idx = end_idx
 
         # remove zero bases and normalize across features
-        H_pre_cluster = H_pre_cluster[np.sum(H_samples, axis=1) != 0.0]
+        H_pre_cluster = H_pre_cluster[np.sum(H_pre_cluster, axis=1) != 0.0]
         H_pre_cluster = normalize(H_pre_cluster, norm='l2', axis=1)
 
         # cluster all bases
         self._logger.info("clustering bases samples")
-        labels = self.cluster.fit_predict(H_samples)
+        labels = self.cluster.fit_predict(H_pre_cluster)
 
         # compute consensus bases from clusters
         cluster_ids = np.unique(labels[labels != -1])
@@ -216,7 +221,8 @@ class UoI_NMF_Base(AbstractDecompositionModel):
         H_cons = np.zeros((n_clusters, n_features))
 
         for c_id in cluster_ids:
-            H_cons[c_id, :] = self.cons_meth(H_samples[labels == c_id], axis=0)
+            H_cons[c_id, :] = self.cons_meth(H_pre_cluster[labels == c_id],
+                                             axis=0)
 
         # re-normalize across features
         H_cons = normalize(H_cons, norm='l2', axis=1)
@@ -396,7 +402,7 @@ class UoI_NMF(UoI_NMF_Base):
         nmf_tol=0.0001, nmf_max_iter=400,
         db_eps=0.5, db_min_samples=None, db_metric='euclidean',
         db_metric_params=None, db_algorithm='auto', db_leaf_size=30,
-        random_state=None, logger=None,
+        use_dissimilarity=True, random_state=None, logger=None,
     ):
         # create NMF solver
         nmf = skNMF(init=nmf_init,
@@ -422,4 +428,5 @@ class UoI_NMF(UoI_NMF_Base):
             nnreg=None,
             cons_meth=np.median,
             random_state=random_state,
+            use_dissimilarity=use_dissimilarity,
             logger=logger)

--- a/pyuoi/decomposition/NMF.py
+++ b/pyuoi/decomposition/NMF.py
@@ -96,7 +96,7 @@ class UoI_NMF_Base(AbstractDecompositionModel):
             if isinstance(ranks, int):
                 self.ranks = list(range(2, ranks + 1)) \
                     if isinstance(ranks, int) else list(ranks)
-            elif isinstance(ranks, (list, tuple, range, np.array)):
+            elif isinstance(ranks, (list, tuple, range, np.ndarray)):
                 self.ranks = tuple(ranks)
             else:
                 raise ValueError('Specify a max value or an array-like for k.')
@@ -312,9 +312,9 @@ class UoI_NMF_Base(AbstractDecompositionModel):
             Data matrix of original shape.
         """
         check_is_fitted(self, ['components_'])
-        n_samples = W.shape[1]
+        n_components = W.shape[1]
 
-        if n_samples != self.components_.shape[0]:
+        if n_components != self.components_.shape[0]:
             raise ValueError(
                 'Incompatible shape: cannot multiply %s with %s.'
                 % (W.shape, self.components_.shape))

--- a/pyuoi/decomposition/NMF.py
+++ b/pyuoi/decomposition/NMF.py
@@ -8,6 +8,7 @@ from .utils import dissimilarity
 from ..utils import check_logger
 
 from itertools import combinations
+from sklearn.base import BaseEstimator
 from sklearn.decomposition import NMF as skNMF
 from sklearn.cluster import DBSCAN
 from sklearn.preprocessing import normalize
@@ -59,6 +60,7 @@ class UoI_NMF_Base(AbstractDecompositionModel):
         self, n_boots=10, ranks=None, nmf=None, cluster=None, nnreg=None,
         cons_meth=None, use_dissimilarity=True, random_state=None, logger=None
     ):
+        super(UoI_NMF_Base, self).__init__()
         self.__initialize(
             n_boots=n_boots,
             ranks=ranks,
@@ -72,7 +74,8 @@ class UoI_NMF_Base(AbstractDecompositionModel):
 
     def set_params(self, **kwargs):
         """Set the parameters of this estimator."""
-        self.__initialize(**kwargs)
+        super(UoI_NMF_Base, self).set_params(**kwargs)
+        self.__initialize(**self.get_params())
 
     def __initialize(self, **kwargs):
         """Initializes the NMF class."""
@@ -119,8 +122,10 @@ class UoI_NMF_Base(AbstractDecompositionModel):
         if nnreg is None:
             self.nnreg = lambda A, b: spo.nnls(A, b)[0]
         else:
-            if isinstance(nnreg, ):
+            if isinstance(nnreg, BaseEstimator):
                 self.nnreg = lambda A, B: nnreg.fit(A, B).coef_
+            elif callable(nnreg):
+                self.nnreg = nnreg
             else:
                 raise ValueError("Unrecognized regressor.")
 
@@ -191,8 +196,6 @@ class UoI_NMF_Base(AbstractDecompositionModel):
                 for boot1, boot2 in combinations(range(self.n_boots), 2):
                     gamma[k_idx] += dissimilarity(H_k[boot1], H_k[boot2])
             k_min = self.ranks[np.argmin(gamma)]
-            print(gamma)
-            print(k_min)
             H_pre_cluster = H_samples[k_min].reshape((self.n_boots * k_min,
                                                       n_features))
         else:
@@ -403,6 +406,7 @@ class UoI_NMF(UoI_NMF_Base):
         db_eps=0.5, db_min_samples=None, db_metric='euclidean',
         db_metric_params=None, db_algorithm='auto', db_leaf_size=30,
         use_dissimilarity=True, random_state=None, logger=None,
+        nmf=None, cluster=None, nnreg=None, cons_meth=None
     ):
         # create NMF solver
         nmf = skNMF(init=nmf_init,

--- a/pyuoi/decomposition/utils.py
+++ b/pyuoi/decomposition/utils.py
@@ -88,3 +88,12 @@ def stability_selection_to_threshold(stability_selection, n_boots):
                          "the correct bounds.")
 
     return selection_threshold
+
+
+def dissimilarity(H1, H2):
+    k = H1.shape[0]
+    H1 = H1 / np.linalg.norm(H1, axis=1, keepdims=True)
+    H2 = H2 / np.linalg.norm(H1, axis=1, keepdims=True)
+    C = np.dot(H1, H2.T)
+    diss = 1 - (np.max(C, axis=0).sum() + np.max(C, axis=1).sum()) / (2. * k)
+    return diss

--- a/pyuoi/decomposition/utils.py
+++ b/pyuoi/decomposition/utils.py
@@ -91,9 +91,24 @@ def stability_selection_to_threshold(stability_selection, n_boots):
 
 
 def dissimilarity(H1, H2):
+    """Calculates the dissimilarity between two sets of NMF bases.
+
+    Parameters
+    ----------
+    H1 : ndarray, shape (n_components, n_features)
+        First set of bases.
+
+    H2 : ndarray, shape (n_components, n_features)
+        Second set of bases.
+
+    Returns
+    -------
+    diss : float
+        Dissimilarity between the two sets of bases.
+    """
     k = H1.shape[0]
     H1 = H1 / np.linalg.norm(H1, axis=1, keepdims=True)
-    H2 = H2 / np.linalg.norm(H1, axis=1, keepdims=True)
+    H2 = H2 / np.linalg.norm(H2, axis=1, keepdims=True)
     C = np.dot(H1, H2.T)
-    diss = 1 - (np.max(C, axis=0).sum() + np.max(C, axis=1).sum()) / (2. * k)
+    diss = 1 - ((np.max(C, axis=0).sum() + np.max(C, axis=1).sum()) / (2. * k))
     return diss

--- a/tests/test_nmf.py
+++ b/tests/test_nmf.py
@@ -3,6 +3,34 @@ import numpy as np
 
 from numpy.testing import assert_array_equal
 from pyuoi.decomposition import UoI_NMF
+from pyuoi.decomposition.utils import dissimilarity
+
+
+def setup():
+    W = np.random.randint(0, high=2, size=(500, 5))
+    H = np.random.randint(0, high=2, size=(5, 2))
+    X = np.dot(W, H)
+    noise = np.random.normal(loc=0, scale=0.5, size=X.shape)**2
+    X = X + noise
+
+    n_boots = 2
+    ranks = 5
+    uoi = UoI_NMF(n_boots=n_boots,
+                  ranks=[ranks],
+                  nmf_max_iter=1000,
+                  random_state=2332)
+    return uoi, X
+
+
+def test_dissimilarity():
+    """Test the dissimilarity function."""
+    k = 5
+    n_features = 20
+
+    # same bases should be a dissimilarity of zero
+    H1 = np.random.randint(low=0, high=3, size=(k, n_features))
+    H2 = np.copy(H1)
+    assert np.allclose(dissimilarity(H1, H2), 0.)
 
 
 @pytest.mark.fast
@@ -11,25 +39,10 @@ def test_UoI_NMF_initialization():
     n_boots = 30
     ranks = 10
     uoi = UoI_NMF(n_boots=n_boots, ranks=ranks)
-
     assert_array_equal(uoi.ranks, np.arange(2, ranks + 1))
     assert uoi.nmf.solver == 'mu'
     assert uoi.nmf.beta_loss == 'kullback-leibler'
     assert uoi.cluster.min_samples == n_boots / 2
-
-
-def setup():
-    W = np.random.randint(0, high=2, size=(500, 5))
-    H = np.random.randint(0, high=2, size=(5, 2))
-    X = np.dot(W, H)
-
-    n_boots = 1
-    ranks = 5
-    uoi = UoI_NMF(n_boots=n_boots,
-                  ranks=[ranks],
-                  nmf_max_iter=1000,
-                  random_state=2332)
-    return uoi, X
 
 
 @pytest.mark.fast
@@ -41,8 +54,18 @@ def test_UoI_NMF_fit():
 
 
 @pytest.mark.fast
-def test_UoI_NMF_tranform():
-    """Tests that the """
+def test_UoI_NMF_fit_no_dissimilarity():
+    """Tests that the fitting procedure of UoI NMF runs without error, when
+    the algorithm does not use dissimilarity to choose a rank."""
+    uoi, X = setup()
+    uoi.set_params(use_dissimilarity=False)
+    uoi.fit(X)
+    assert hasattr(uoi, 'components_')
+
+
+@pytest.mark.fast
+def test_UoI_NMF_transform():
+    """Tests that the transform procedure of UoI NMF runs without error."""
     uoi, X = setup()
     X_tfm = uoi.fit_transform(X)
     assert hasattr(uoi, 'components_')
@@ -50,18 +73,9 @@ def test_UoI_NMF_tranform():
 
 
 @pytest.mark.fast
-def test_UoI_NMF_fit_tranform():
-    """Tests that the """
-    uoi, X = setup()
-    uoi.fit(X)
-    X_tfm = uoi.transform(X)
-    assert hasattr(uoi, 'components_')
-    assert X_tfm is not None
-
-
-@pytest.mark.fast
 def test_UoI_NMF_reconstruction_error():
-    """Tests that the """
+    """Tests that a reconstruction error is calculated when data is
+    transformed."""
     uoi, X = setup()
     uoi.fit(X)
     X_tfm = uoi.transform(X, reconstruction_err=True)

--- a/tests/test_nmf.py
+++ b/tests/test_nmf.py
@@ -1,25 +1,21 @@
 import pytest
 import numpy as np
 
-from numpy.testing import assert_array_equal
-from pyuoi.decomposition import UoI_NMF
+from numpy.testing import assert_array_equal, assert_raises
+from pyuoi.decomposition import UoI_NMF, UoI_NMF_Base
 from pyuoi.decomposition.utils import dissimilarity
+from sklearn.cluster import DBSCAN
+from sklearn.decomposition import NMF
 
 
-def setup():
+@pytest.fixture
+def nmf_setup():
     W = np.random.randint(0, high=2, size=(500, 5))
     H = np.random.randint(0, high=2, size=(5, 2))
     X = np.dot(W, H)
     noise = np.random.normal(loc=0, scale=0.5, size=X.shape)**2
     X = X + noise
-
-    n_boots = 2
-    ranks = 5
-    uoi = UoI_NMF(n_boots=n_boots,
-                  ranks=[ranks],
-                  nmf_max_iter=1000,
-                  random_state=2332)
-    return uoi, X
+    return X
 
 
 def test_dissimilarity():
@@ -34,6 +30,19 @@ def test_dissimilarity():
 
 
 @pytest.mark.fast
+def test_UoI_NMF_Base_initialization():
+    """Tests the initialization of UoI NMF Base Class."""
+    n_boots = 30
+    ranks = 10
+    uoi = UoI_NMF(n_boots=n_boots, ranks=ranks,
+                  random_state=np.random.RandomState(2332))
+    assert_array_equal(uoi.ranks, np.arange(2, ranks + 1))
+    assert uoi.nmf.solver == 'mu'
+    assert uoi.nmf.beta_loss == 'kullback-leibler'
+    assert uoi.cluster.min_samples == n_boots / 2
+
+
+@pytest.mark.fast
 def test_UoI_NMF_initialization():
     """Tests the initialization of UoI NMF."""
     n_boots = 30
@@ -43,40 +52,98 @@ def test_UoI_NMF_initialization():
     assert uoi.nmf.solver == 'mu'
     assert uoi.nmf.beta_loss == 'kullback-leibler'
     assert uoi.cluster.min_samples == n_boots / 2
+    assert uoi.cons_meth == np.median
 
 
 @pytest.mark.fast
-def test_UoI_NMF_fit():
+def test_UoI_NMF_initialization_value_error():
+    """Tests that ValueErrors are correctly raised in the NMF initialization."""
+    assert_raises(ValueError, UoI_NMF_Base, **{'ranks': 2.5})
+    assert_raises(ValueError, UoI_NMF_Base, **{'nmf': NMF})
+    assert_raises(ValueError, UoI_NMF_Base, **{'cluster': DBSCAN})
+    assert_raises(ValueError, UoI_NMF_Base, **{'nnreg': 2})
+
+
+@pytest.mark.fast
+def test_UoI_NMF_fit(nmf_setup):
     """Tests that the fitting procedure of UoI NMF runs without error."""
-    uoi, X = setup()
+    X = nmf_setup
+
+    n_boots = 1
+    ranks = 5
+    uoi = UoI_NMF(n_boots=n_boots,
+                  ranks=[ranks],
+                  nmf_max_iter=1000,
+                  random_state=2332)
     uoi.fit(X)
     assert hasattr(uoi, 'components_')
 
 
 @pytest.mark.fast
-def test_UoI_NMF_fit_no_dissimilarity():
+def test_UoI_NMF_fit_no_dissimilarity(nmf_setup):
     """Tests that the fitting procedure of UoI NMF runs without error, when
     the algorithm does not use dissimilarity to choose a rank."""
-    uoi, X = setup()
+    X = nmf_setup
+
+    n_boots = 1
+    ranks = 5
+    uoi = UoI_NMF(n_boots=n_boots,
+                  ranks=[ranks],
+                  nmf_max_iter=1000,
+                  random_state=2332)
     uoi.set_params(use_dissimilarity=False)
     uoi.fit(X)
     assert hasattr(uoi, 'components_')
 
 
 @pytest.mark.fast
-def test_UoI_NMF_transform():
+def test_UoI_NMF_transform(nmf_setup):
     """Tests that the transform procedure of UoI NMF runs without error."""
-    uoi, X = setup()
+    X = nmf_setup
+
+    n_boots = 1
+    ranks = 5
+    uoi = UoI_NMF(n_boots=n_boots,
+                  ranks=[ranks],
+                  nmf_max_iter=1000,
+                  random_state=2332)
     X_tfm = uoi.fit_transform(X)
     assert hasattr(uoi, 'components_')
     assert X_tfm is not None
 
 
 @pytest.mark.fast
-def test_UoI_NMF_reconstruction_error():
+def test_UoI_NMF_transform_value_error(nmf_setup):
+    """Tests that the transform procedure of UoI NMF correctly raises a
+    ValueError."""
+    X = nmf_setup
+    n_boots = 1
+    ranks = 5
+    uoi = UoI_NMF(n_boots=n_boots,
+                  ranks=[ranks],
+                  nmf_max_iter=1000,
+                  random_state=2332)
+    uoi.fit(X)
+
+    # transform
+    Y = np.random.normal(size=(X.shape[0], 2 * X.shape[1]))**2
+    assert_raises(ValueError, uoi.transform, Y)
+    # inverse transform
+    W = np.random.normal(size=(X.shape[0], 2 * uoi.components_.shape[0]))**2
+    assert_raises(ValueError, uoi.inverse_transform, W)
+
+
+@pytest.mark.fast
+def test_UoI_NMF_reconstruction_error(nmf_setup):
     """Tests that a reconstruction error is calculated when data is
     transformed."""
-    uoi, X = setup()
+    X = nmf_setup
+    n_boots = 1
+    ranks = 5
+    uoi = UoI_NMF(n_boots=n_boots,
+                  ranks=[ranks],
+                  nmf_max_iter=1000,
+                  random_state=2332)
     uoi.fit(X)
     X_tfm = uoi.transform(X, reconstruction_err=True)
     assert hasattr(uoi, 'components_')
@@ -102,7 +169,7 @@ def test_UoI_NMF_correct_number_of_components():
     # fit uoi nmf
     uoi = UoI_NMF(n_boots=10,
                   ranks=[2, 4, 8],
-                  nmf_max_iter=1000,
+                  nmf_max_iter=5000,
                   use_dissimilarity=True)
     uoi.fit(A)
 

--- a/tests/test_nmf.py
+++ b/tests/test_nmf.py
@@ -83,3 +83,27 @@ def test_UoI_NMF_reconstruction_error():
     assert hasattr(uoi, 'reconstruction_err_')
     assert uoi.reconstruction_err_ is not None
     assert X_tfm is not None
+
+
+@pytest.mark.slow
+def test_UoI_NMF_correct_number_of_components():
+    """Tests that, using the dissimilarity metric, UoI NMF extracts the correct
+    number of bases."""
+    k = 2
+    n_samples = 1000
+    n_features = 30
+
+    # create data matrix
+    W = np.random.randint(low=0, high=3, size=(n_samples, k))
+    H = np.random.randint(low=0, high=3, size=(k, n_features))
+    noise = np.random.normal(loc=0, scale=0.5, size=(n_samples, n_features))**2
+    A = np.dot(W, H) + noise
+
+    # fit uoi nmf
+    uoi = UoI_NMF(n_boots=10,
+                  ranks=[2, 4, 8],
+                  nmf_max_iter=1000,
+                  use_dissimilarity=True)
+    uoi.fit(A)
+
+    assert uoi.components_.shape[0] == k

